flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/stats/TaskStatsRequestCoordinator.java

```diff
@@ -0,0 +1,323 @@
+     */
+    public void shutDown() {
+        synchronized (lock) {
+            if (!isShutDown) {
+                log.info("Shutting down task stats request coordinator.");
+
+                for (PendingStatsRequest<T, V> pending : pendingRequests.values()) {
+                    pending.discard(new RuntimeException("Shut down"));
+                }
+
+                pendingRequests.clear();
+                recentPendingRequestIds.clear();
+
+                isShutDown = true;
+            }
+        }
+    }
+
+    /**
+     * Handles the successfully returned task stats response by collecting the corresponding subtask
+     * samples.
+     *
+     * @param requestId ID of the request.
+     * @param executionId ID of the sampled task.
+     * @param result Result of stats request returned by an individual task.
+     * @throws IllegalStateException If unknown request ID and not recently finished or cancelled
+     *     sample.
+     */
+    public void handleSuccessfulResponse(int requestId, ExecutionAttemptID executionId, T result) {
+
+        synchronized (lock) {
+            if (isShutDown) {
+                return;
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug("Collecting stats sample {} of task {}", requestId, executionId);
+            }
+
+            PendingStatsRequest<T, V> pending = pendingRequests.get(requestId);
+
+            if (pending != null) {
+                pending.collectTaskStats(executionId, result);
+
+                // Publish the sample
+                if (pending.isComplete()) {
+                    pendingRequests.remove(requestId);
+                    rememberRecentRequestId(requestId);
+
+                    pending.completePromiseAndDiscard();
+                }
+            } else if (recentPendingRequestIds.contains(requestId)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Received late stats sample {} of task {}", requestId, executionId);
+                }
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("Unknown request ID %d.", requestId));
+                }
+            }
+        }
+    }
+
+    private void rememberRecentRequestId(int sampleId) {
+        if (recentPendingRequestIds.size() >= numGhostSampleIds) {
+            recentPendingRequestIds.removeFirst();
+        }
+        recentPendingRequestIds.addLast(sampleId);
+    }
+
+    @VisibleForTesting
+    public int getNumberOfPendingRequests() {
+        synchronized (lock) {
+            return pendingRequests.size();
+        }
+    }
+
+    // ------------------------------------------------------------------------
+
+    /**
+     * A pending task stats request, which collects samples from individual tasks and completes the
+     * response future upon gathering all of of them.
+     *
+     * <p>Has to be accessed in lock scope.
```

Comment:
Might be easier to add `@NotThreadSafe`.

Reply:
added the annotation.